### PR TITLE
[SPARK-17758][SQL] Last returns wrong result in case of empty partition

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

The result of the `Last` function can be wrong when the last partition processed is empty. It can return `null` instead of the expected value. For example, this can happen when we process partitions in the following order:

```
- Partition 1 [Row1, Row2]
- Partition 2 [Row3]
- Partition 3 []
```

In this case the `Last` function will currently return a null, instead of the value of `Row3`.

This PR fixes this by adding a `valueSet` flag to the `Last` function.
## How was this patch tested?

We only used end to end tests for `DeclarativeAggregateFunction`s. I have added an evaluator for these functions so we can tests them in catalyst. I have added a `LastTestSuite` to test the `Last` aggregate function.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/LastTestSuite.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    // Update - Merge - Eval
+    val p2 = evaluator.update(InternalRow(2), InternalRow(10))
+    val m1 = evaluator.merge(p1, p0, p2)
+    assert(evaluator.eval(m1) === InternalRow(10))
+
+    // Update - Merge - Eval (empty partition at the end)
+    val m2 = evaluator.merge(p2, p1, p0)
+    assert(evaluator.eval(m2) === InternalRow(-99))
+  }
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/LastTestSuite.scala]: Have a test to call update using a null input and then check the answer?

**Comment by yhuai** [comment]
Thank you for fixing this! It is great to have unit tests to test individual aggregate functions. We can start to add more tests for other functions.

**Comment by davies** [comment]
Spark does not preserve the order of partitions, how can we make sure the result is Row3, rather than Row2?

**Comment by hvanhovell** [comment]
You cannot. Last is not deterministic outside of ordered window functions. The only thing is that you can have inconsistent results if the last physical partition happens to be an empty one.

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 2.0.

**Pull request closed without merging**